radixdlt-java/src/main/java/com/radixdlt/client/core/network/RadixNetworkController.java

```diff
@@ -0,0 +1,143 @@
+ * to produce a stream of actions and states.
+ */
+public class RadixNetworkController implements AtomSubmitter {
+	private static final Logger LOGGER = LoggerFactory.getLogger(RadixNetworkController.class);
+
+	public static class RadixNetworkControllerBuilder {
+		private RadixNetwork network;
+		private List<RadixNetworkEpic> epics = new ArrayList<>();
+
+		public RadixNetworkControllerBuilder() {
+		}
+
+		public RadixNetworkControllerBuilder network(RadixNetwork network) {
+			this.network = network;
+			return this;
+		}
+
+		public RadixNetworkControllerBuilder addEpic(RadixNetworkEpic epic) {
+			this.epics.add(epic);
+			return this;
+		}
+
+		public RadixNetworkController build() {
+			Objects.requireNonNull(network);
+
+			return new RadixNetworkController(network, epics);
+		}
+	}
+
+	private final BehaviorSubject<RadixNetworkState> networkState;
+
+	private final Subject<RadixNodeAction> nodeActions = PublishSubject.<RadixNodeAction>create().toSerialized();
+
+	private RadixNetworkController(RadixNetwork network, List<RadixNetworkEpic> epics) {
+		this.networkState = BehaviorSubject.createDefault(new RadixNetworkState(Collections.emptyMap()));
+
+		// Run reducers first
+		ConnectableObservable<RadixNodeAction> reducedNodeActions = nodeActions.doOnNext(action -> {
+			LOGGER.info("NEXT ACTION: " + action.toString());
+			RadixNetworkState nextState = network.reduce(networkState.getValue(), action);
+
+			// TODO: remove null check and just check for equality
+			if (nextState != null) {
+				networkState.onNext(nextState);
+			}
+
+			// TODO: turn this into an action/state pair so synchronized
+		}).publish();
+
+		// Then run Epics
+		Set<Observable<RadixNodeAction>> updates = epics.stream()
+			.map(epic -> epic.epic(reducedNodeActions, networkState))
+			.collect(Collectors.toSet());
+		Observable.merge(updates).subscribe(this::dispatch);
+
+		reducedNodeActions.connect();
+	}
+
+	public Observable<RadixNetworkState> getNetwork() {
+		return networkState;
+	}
+
+	public Observable<RadixNodeAction> getActions() {
+		return nodeActions;
+	}
+
+	public void dispatch(RadixNodeAction action) {
+		nodeActions.onNext(action);
+	}
+
+	/**
+	 * Immediately submits an atom into the ledger without waiting for subscription. The returned
+	 * observable is a full replay of the status of the atom, from submission to acceptance by
+	 * the network.
+	 *
+	 * @param atom atom to submit into the ledger
+	 * @return Observable emitting status updates to submission
+	 */
+	@Override
+	public Observable<SubmitAtomAction> submitAtom(Atom atom) {
+		SubmitAtomAction initialAction = SubmitAtomRequestAction.newRequest(atom);
+
+		Observable<SubmitAtomAction> status =
+			nodeActions.ofType(SubmitAtomAction.class)
+				.filter(u -> u.getUuid().equals(initialAction.getUuid()))
+				.takeUntil(u -> u instanceof SubmitAtomResultAction);
+		ConnectableObservable<SubmitAtomAction> replay = status.replay();
```

Comment:
This is an artifact from the earlier api where it was weird to have a `submitAtom` method which didn't do anything until one called `subscribe` to the result of the method. The replay was added so that this observable can connect immediately with the results being saved for a subscriber to view without losing information.

Having said that, this will be refactored out later on now that there is a generic `dispatch` and `getActions` method in the controller.